src/maker_protocol.rs

```diff
@@ -205,72 +224,119 @@ fn handle_message(
-        None => NEWLY_CONNECTED_TAKER_ALLOWED_METHODS
-            .iter()
-            .any(|&r| r == method),
-    };
-    if !is_method_allowed {
-        return Err(Error::Protocol("unexpected method"));
-    }
-
     let request: TakerToMakerMessage = match serde_json::from_str(&line) {
         Ok(r) => r,
         Err(_e) => return Err(Error::Protocol("message parsing error")),
     };
     log::trace!(target: "maker", "<== {:?}", request);
-    let outgoing_message = match request {
-        TakerToMakerMessage::TakerHello(_message) => {
-            connection_state.allowed_method = None;
-            None
-        }
-        TakerToMakerMessage::GiveOffer(_message) => {
-            let max_size = wallet.read().unwrap().get_offer_maxsize(rpc)?;
-            let tweakable_point = wallet.read().unwrap().get_tweakable_keypair().1;
-            connection_state.allowed_method = Some("signsenderscontracttx");
-            Some(MakerToTakerMessage::Offer(Offer {
-                absolute_fee: 1000,
-                amount_relative_fee: 0.005,
-                max_size,
-                min_size: 10000,
-                tweakable_point,
-            }))
+
+    let outgoing_message = match connection_state.allowed_message {
+        ExpectedMessage::TakerHello => {
+            if let TakerToMakerMessage::TakerHello(_) = request {
+                connection_state.allowed_message = ExpectedMessage::NewlyConnectedTaker;
+                None
+            } else {
+                return Err(Error::Protocol("Expected Taker Hello Message"));
+            }
         }
-        TakerToMakerMessage::SignSendersContractTx(message) => {
-            connection_state.allowed_method = Some("proofoffunding");
-            handle_sign_senders_contract_tx(wallet, message)?
+
+        ExpectedMessage::NewlyConnectedTaker => match request {
+            TakerToMakerMessage::GiveOffer(_) => {
+                let max_size = wallet.read().unwrap().get_offer_maxsize(rpc)?;
+                let tweakable_point = wallet.read().unwrap().get_tweakable_keypair().1;
+                connection_state.allowed_message = ExpectedMessage::SignSendersContractTx;
+                Some(MakerToTakerMessage::Offer(Offer {
+                    absolute_fee: 1000,
+                    amount_relative_fee: 0.005,
+                    max_size,
+                    min_size: 10000,
+                    tweakable_point,
+                }))
+            }
+
+            TakerToMakerMessage::SignSendersContractTx(message) => {
+                connection_state.allowed_message = ExpectedMessage::ProofOfFunding;
+                handle_sign_senders_contract_tx(wallet, message)?
+            }
+
+            TakerToMakerMessage::ProofOfFunding(proof) => {
+                connection_state.allowed_message = ExpectedMessage::SendersAndReceiversContractSigs;
+                handle_proof_of_funding(connection_state, rpc, wallet, &proof)?
+            }
+
```

Comment:
These newlines don't do much and make less of the code fit on one page. Consider removing them. The same thing happens in the definition of the enum ExpectedMessage